File: src/jef/compiler/renderer/stencil_renderer.rb
# To change this template, choose Tools | Templates
# and open the template in the editor.

require 'renderer/base_renderer'

class StencilRenderer < BaseRenderer
	
	@root
	
  def initialize root
		@output = ""
    @root = root
  end
	
	def render
		render_renderers @root
	end
	
	def add_output str
		super "\t"+str
	end
	
	def output
		@output
	end
	
	def render_html_node node
		render_in = "placeholder";
		if ( node.parent != @root )then
			render_in = varname( node.parent )
		end
		add_output "var "+varname( node )+" = document.createElement( \""+node.tag+"\" );"
		node.attributes.each do |a|
			puts a
			add_output varname( node ) + ".setAttribute( \"" + a.name + "\", " + a.value.js_expression + " );";
		end
		
		render_renderers node
		add_output render_in+".appendChild( "+varname( node )+" );"
		
	end
	
	def render_component_node node
		render_in = "placeholder"
		if ( node.parent != @root )then
			render_in = varname( node.parent )
		end
		add_output "this._templateControls[\""+varname( node )+"\"].renderContentsInPlaceholder( "+render_in+" );"
		
	end
	
	def render_text_node node
		render_in = "placeholder";
		if ( node.parent != @root )then
			render_in = varname( node.parent )
		end
		add_output "var "+varname( node )+" = document.createTextNode( "+node.get_js_expression+" );"
		add_output render_in+".appendChild( "+varname( node )+" );"
	end
	
	def render_renderers node
		node.children.each do |n|
			
			if ( n.instance_of? HtmlNode ) then
				render_html_node n
			elsif ( n.instance_of? ComponentNode ) then
				render_component_node n
			elsif ( n.instance_of? StencilNode ) then
			#	render_stencil_node n
			elsif ( n.instance_of? TextNode ) then
				render_text_node n
			else
				render_initializers n
			end
			
		end
	end

end

File: compiler/renderer/StencilRenderer.ts
import BaseRenderer from "./BaseRenderer";
import TemplateNode from "../template/TemplateNode";
import HtmlNode from "../template/HtmlNode";
import ComponentNode from "../template/ComponentNode";
import TextNode from "../template/TextNode";
import StencilNode from "../template/StencilNode";

import Attribute from '../template/Attribute';

export default class StencilRenderer extends BaseRenderer {

	root:TemplateNode;

	constructor(node:TemplateNode, debug: number, language: string ){
		super( debug, language );
		this.root = node;
	}

	render(){
		this.renderRenderers(this.root);
	}

	renderRenderers(node:TemplateNode){
		for (let i = 0; i < node.children.length; i++ ){
			let child = node.children[i];

			if ( child instanceof HtmlNode ){
				this.renderHtmlNode(child);
			}
			else if ( child instanceof ComponentNode ){
				this.renderComponentNode(child);
			}
			else if ( child instanceof StencilNode ){
				//skip
			}
			else if ( child instanceof TextNode ){
				this.renderTextNode(child);
			}
			else {
				//render_initializers?
			}

		}
	}

	renderHtmlNode( node:HtmlNode ){
		var render_in = "placeholder";
		if ( node.parent !== this.root ){
			render_in = this.getVarname( node.parent );
		}

		if ( node.namespace !== null ){
			this.addOutput( "var " + this.getVarname(node) + " = document.createElementNS( \"" + node.namespace + "\", \"" + node.tag + "\" );" );
		} else {
			this.addOutput( "var " + this.getVarname(node) + " = document.createElement( \"" + node.tag + "\" );" );
		}

		for (let i = 0; i < node.attributes.length; i++ ){
			let attribute:Attribute = node.attributes[i];

			if ( attribute.namespace == null ){
				if ( attribute.name === "style" ){
					this.addOutput( this.getVarname(node)+".style.cssText = "+attribute.value.getExpression( this.language )+";" );
				}else{
					this.addOutput( this.getVarname(node) + ".setAttribute( \""+attribute.name+"\", "+attribute.value.getExpression( this.language )+" );" );
				}
			}else{
				this.addOutput( this.getVarname(node) + ".setAttributeNS( \""+attribute.namespace+"\", \""+attribute.name+"\", "+attribute.value.getExpression( this.language )+" );" );
			}
		}

		this.renderRenderers(node);

		this.addOutput(render_in + ".appendChild( " +this.getVarname(node)+ " );");

	}

	renderComponentNode( node:ComponentNode ){
		var render_in = "placeholder";
		if ( node.parent !== this.root ){
			render_in = this.getVarname( node.parent );
		}
		this.addOutput( "this._templateControls[\""+this.getVarname( node )+"\"].renderContentsInPlaceholder( "+render_in+" );" );
	}

	renderTextNode( node:TextNode ){
		 if ( ! node.getStripSpaces()
		 	|| node.raw.trim().length > 0 ){

			var render_in = "placeholder";
			if ( node.parent !== this.root ){
				render_in = this.getVarname( node.parent );
			}

			this.addOutput( "var "+this.getVarname(node)+" = document.createTextNode( "+node.expression.getExpression( this.language )+" );" );
			this.addOutput( render_in + ".appendChild( " + this.getVarname(node) + " );" );
		 }
	}

}